mesh/spacing.hpp

```diff
@@ -0,0 +1,689 @@
+             } SPACING_TYPE;
+
+/// Class for spacing functions that define meshes in a dimension, using a
+/// formula or method implemented in a derived class.
+class SpacingFunction
+{
+public:
+   /** @brief Base class constructor.
+   @param[in] n_  Size or number of intervals, which defines elements.
+   @param[in] r   Whether to reverse the spacings, false by default.
+   @param[in] s   Whether to scale parameters by the refinement or coarsening
+                  factor, in the function @a SpacingFunction::ScaleParameters.
+   */
+   SpacingFunction(int n_, bool r=false, bool s=false)
+   {
+      n = n_;
+      reverse = r;
+      scale = s;
+   }
+
+   /// Returns the size, or number of intervals (elements).
+   inline int Size() const { return n; }
+
+   /// Sets the size, or number of intervals (elements).
+   virtual void SetSize(int size) = 0;
+
+   /// Sets the property that determines whether the spacing is reversed.
+   void SetReverse(bool r) { reverse = r; }
+
+   /// Returns the width of interval @a p (between 0 and @a Size() - 1).
+   virtual double Eval(int p) = 0;
+
+   /// Returns the width of all intervals, resizing @a s to @a Size().
+   void EvalAll(Vector & s)
+   {
+      s.SetSize(n);
+      for (int i=0; i<n; ++i)
+      {
+         s[i] = Eval(i);
+      }
+   }
+
+   /** @brief Scales parameters by the factor @a a associated with @a Size().
+
+       Note that parameters may be scaled inversely during coarsening and
+       refining, so the scaling should be linear in the sense that scaling by a
+       number followed by scaling by its inverse has no effect on parameters. */
+   virtual void ScaleParameters(double a) { }
+
+   /// Returns the spacing type, indicating the derived class.
+   virtual SPACING_TYPE SpacingType() const = 0;
+
+   /** @brief Prints all the data necessary to define the spacing function and
+       its current state (size and other parameters).
+
+       The format is generally
+       SPACING_TYPE numIntParam numDoubleParam {int params} {double params} */
+   virtual void Print(std::ostream &os) const = 0;
+
+   /// Returns the number of integer parameters defining the spacing function.
+   virtual int NumIntParameters() const = 0;
+
+   /// Returns the number of double parameters defining the spacing function.
+   virtual int NumDoubleParameters() const = 0;
+
+   /// Returns the array of integer parameters defining the spacing function.
+   /// @param[out] p  Array of integer parameters, resized appropriately.
+   virtual void GetIntParameters(Array<int> & p) const = 0;
+
+   /// Returns the array of double parameters defining the spacing function.
+   /// @param[out] p  Array of double parameters, resized appropriately.
+   virtual void GetDoubleParameters(Vector & p) const = 0;
+
+   /// Returns true if the spacing function is nested during refinement.
+   virtual bool Nested() const = 0;
+
+   /// Returns a clone of this spacing function.
+   virtual SpacingFunction *Clone() const;
```

Comment:
Cut this off at the source, have these return `std::unique_ptr<SpacingFunction>`: a unique ptr can be assigned to a `shared_ptr` at any moment and the underlying resources be placed under management. This can be done by changing the `Clone` method in all the children to have `std::unique_ptr<Child>(new Child(....));` then everything else should be the same, but without the risk of callers needing to remember to use `delete`.

The only complicated `Clone` I can see is `PiecewiseSpacingFunction`, which suggests it's doing something more complex than just copying the underlying type. Why can't that just use the copy ctor?

Reply:
There were two things: 

1) `Clone` should return a `unique_ptr`, so that users are not responsible for managing lifetime. I think this is a good idea but looking at `Array` restriction to trivial types, would mean using another container like `std::vector` for the `pieces` member within `PiecewiseSpacingFunction` and elsewhere if stacking a lot of these `Clone`'d pointers. This does result in changes to [how registers are used](https://stackoverflow.com/questions/58339165/why-can-a-t-be-passed-in-register-but-a-unique-ptrt-cannot) but the reduction in memory risk of using unique ptr compared to raw is easily worth it. 

and 2) for the `Clone`, piecewise should be implemented in terms of the copy constructor like for the rest.
```
   std::unique_ptr<SpacingFunction> Clone() const override
   {
      return std::unique_ptr<PiecewiseSpacingFunction>(new PiecewiseSpacingFunction(*this));
   }
```
or if you decide against `unique_ptr`
```
   SpacingFunction* Clone() const override
   {
      return new PiecewiseSpacingFunction(*this);
   }
```

For `PiecewiseSpacingFunction` the special member functions should then be filled in
```
   PiecewiseSpacingFunction(const PiecewiseSpacingFunction &sf)
   : SpacingFunction(sf),
   np(sf.np), partition(sf.partition), npartition(sf.npartition), pieces(), n0(sf.n0), s(sf.s)
   {
      // To copy, the pointers must be cloned.
      for (const auto &f : sf.pieces) { pieces.Append(f->Clone()); }
   }
   PiecewiseSpacingFunction& operator=(const PiecewiseSpacingFunction &sf)
   {
      PiecewiseSpacingFunction tmp(sf);
      std::swap(tmp, *this);
      return *this;
   }
   PiecewiseSpacingFunction(PiecewiseSpacingFunction &&sf) = default;
   PiecewiseSpacingFunction& operator=(PiecewiseSpacingFunction &&sf) = default;
```
The copy constructor is virtually identical to a vanilla copy ctor, but the notion of "copying" the pointers has to be handled manually to avoid muddying the notion of ownership of the pointers. The moves can just be default as the pointers can be stolen.

This then ensures that `PiecewiseSpacingFunction` doesn't end up with different value semantics compared to the other classes for whom a vanilla copy constructor is fine (no pointers to deep copy). It also means that all the Clone methods look the same, and keep move semantics.

Reply:
Converting from `unique_ptr` to `shared_ptr` is exactly as the stl designers intended it to be used. A pointer starts life with a single owner under unique, but once ownership gets muddied the conversion to shared is totally painless as you found. For completeness the other `Clone` method should use the copy constructors too `(*this);`, as the notion of cloning here is virtual copy construction and this expresses that intent.

Reply:
The other `Clone` methods are updated now.